traffic_ops/traffic_ops_golang/test/error_context.go

```diff
@@ -0,0 +1,267 @@
+// panic must be turned on before errors are created
+// once turned on the panic mode can't be turned off
+func (ec *ErrorContext) TurnPanicOn() error {
+	if ec.calledNewError {
+		return ec.internalError(BAD_INIT_TIMING, nil)
+	}
+	ec.doPanic = true
+	return nil
+}
+
+// AddMapping gives a default message for a given error code.
+// Mappings should be configured before errors are called for consistency.
+//
+// usage: ErrorContext.AddMapping(404, "not found")
+//        err := ErrorContext.NewError(404)
+//        err := ErrorContext.NewError(404, "not found: %v", prev_err")
+// the err automatically has a default message
+// the default message can be overridden with something else to add context to the error
+//
+// parameters:
+//  `code` should exist in the error context, only one mapping can exist per error context
+//  `msg` should be a plain string without special formatting
+func (ec *ErrorContext) AddMapping(code uint, msg string) error {
+	if ec.calledNewError {
+		return ec.internalError(BAD_INIT_TIMING, nil)
+	}
+	if !ec.whitelisted(code) {
+		return ec.internalError(BAD_MAP_CREATE, code)
+	}
+	if _, exists := ec.description[code]; exists {
+		return ec.internalError(BAD_DUP_MAPPING, code)
+	}
+	ec.description[code] = msg
+	return nil
+}
+
+// AddMap applies the `AddMapping` function for every element in the given map
+// the function argument will not override the current contents of the map, everything is additive
+// an error is returned if a duplicate code is added
+func (ec *ErrorContext) AddMap(mapping map[uint]string) error {
+	for code, desc := range mapping {
+		if err := ec.AddMapping(code, desc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// all internal errors
+const (
+	BAD_ERROR_CODE  = -1 - iota // when bad error code is given in creation of new error
+	BAD_MAP_LOOKUP              // when creating an error with no error message, default message wasn't found
+	BAD_MAP_CREATE              // when bad error code is given in creation of map
+	BAD_DUP_MAPPING             // when a mapping is already made
+	BAD_FMT_STRING              // when the fmt string isn't a string
+	BAD_INIT_TIMING             // when the error context is modifed after having created an error
+)
+
+func (ec ErrorContext) internalError(code int, offender interface{}) ErrorCoder {
+
+	var err error
+	switch code {
+	case BAD_ERROR_CODE:
+		err = fmt.Errorf("error code %v not found in whitelist for \"%s\" error context", offender, ec.name)
+	case BAD_MAP_CREATE:
+		err = fmt.Errorf("when creating default error mapping, code %v wasn't found in the code whitelist for the \"%s\" error context", offender, ec.name)
+	case BAD_DUP_MAPPING:
+		err = fmt.Errorf("when creating default error mapping, code %v was already found as a mapping for the \"%s\" error context", offender, ec.name)
+	case BAD_MAP_LOOKUP:
+		err = fmt.Errorf("bad default error lookup for code %v in the \"%s\" error context", offender, ec.name)
```

Comment:
I'm not really clear on what "offender" is? It looks like it's being used as the code in some errors, but `the leading argument \"%v\"` on others? And it's an `interface{}`? 

Would you mind adding a function-level comment explaining the `offender` parameter?